### PR TITLE
remove cancellation check from retry policies

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: #1064 

Bases on the related issue comments and conclusion, this PR is updating the places where cancellation is checked.

Note how BodyStream is now offering the cancellation checking by default for any Read() operation.
All derived classes can still access the context and check again if they want, but now we are making sure all derived class will always check as the first thing.

### CODE REVIEW

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:52]: We should settle on a naming convention. `Implementation` vs `Internal` vs something else.

Also, if no one other than `BodyStream` is supposed to call this, make it `private`.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:46]:
```suggestion
     * @param context #Context so that operation can be canceled.
```

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:47]: It doesn't have to be the first, right?
```suggestion
     * @param buffer Pointer to a byte buffer to read the data into.
```

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:77]:
```suggestion
     * @remark Checks and throws if context is cancelled. Otherwise, calls the read implementation.
```

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:85]: Why not keep this virtual so someone can override it if they want to, but leave it as **non-pure** virtual?

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:52]: This as a public API surface name, isn't great. We'd want most (if not all) folks to call `Read`, which is a lot more user friendly, and not call `ReadImplementation` directly.

File: sdk/core/azure-core/src/http/body_stream.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/body_stream.cpp:112]: Probably out of scope for this PR:
Should this try to do async i/o rather than sync and honor context cancellation mid-way, using `CancelIoEx`? And possibly use `ReadFileEx` for async read?
https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-readfile#remarks

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:604]: We should keep and move this cancel into the inner loop and before each call to `ReadFromSocket`, so that we don't wait more than one I/O operation to honor the cancelation.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/retry_policy.cpp:186]: Why are we removing this? Consider moving this above the thread sleep line: 
```C++
if (retryAfter.count() > 0)
```

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: I am surprised we are removing these. Please leave them or help me understand why we need to get rid of them.

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
There are some places that I don't understand nor agree with the removal of the context cancellation check. Let's resolve the discussion before merging, hence marking as blocked.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:52]: All folks will call `Read()`.  All folks are currently and will continue to call `Read()` only.

`ReadImplementation()` is a contract for anyone creating a new bodyStream to implement.   
In the past, the contract was to implement `Read()`, but now `Read()` should not be overridden.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:85]: The main intention is to ensure to end-users that no matter what derived-bodyStream-class they use, the bodyStream will check for cancellation as the first instruction when calling `Read()`.
If we make this virtual, someone could remove the cancellation check and break this guarantee.

File: sdk/core/azure-core/src/http/body_stream.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/body_stream.cpp:112]: That's a good improvement we can add in the future if someone ask for it right?
Keep in mind that then the SDK is reading a bodyStream for sending a request, we will call `Read()` multiple times in chunks, so we are already kind of asking if the customer wants to cancel the request while a file is been read.

Adding the cacellation check for FileBody implementation would empower the end-user to cancel the read while the bodyStream is used out of an HTTP request with the SDK

File: sdk/core/azure-core/src/http/curl/curl.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:604]: TransportPolicy checks for cancellation. This is the `don't wait more than one I/O operation honoring`.

Then `ReadFromSocket` checks for cancellation if the I/O can't be done when requested and until the I/O starts 🔥 !

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Based on #1064.

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: Calling `Read()` from a winHttpBodyStream will fisrt call the base body stream `Read` method which makes this cancellation check. Then that method calls this ReadImplementation.  Leaving this check will mean an instant double check for cancellation.

And the removal of the second check below is because we want to do the cancellation check only before I/O calls.  
Calling `WinHttpQueryDataAvailable` is not an I/O because winHTTP will just tell you how much data is there for you without reading any socket/network.
It is not until we call `WinHttpReadData` that we are doing an I/O

Comment (JeffreyRichter) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: Yes, and I think now that ReadImplementation should be renamed OnRead because this OnXxx pattern is common in OOP. You see it in .NET a lot; maybe it's common in C++ too (@CaseyCarter ?)

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (JeffreyRichter) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:52]: I think we should rename ReadImplementation to OnRead as this is a common OOP pattern.
@CaseyCarter: is this OnXxx pattern common in C++?

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Sure, but this wasn't addressed: https://github.com/Azure/azure-sdk-for-cpp/issues/1064#issuecomment-741289676

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: > Calling `WinHttpQueryDataAvailable` is not an I/O because winHTTP will just tell you how much data is there for you without reading any socket/network.
It is not until we call `WinHttpReadData` that we are doing an I/O

Are you sure about that? Given some of the potential error return codes and the fact we allow async here, I do think the underlying implementation might do an I/O here.

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: In that case, we could leave the second check

Comment (vhvb1989) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: In that case, we could leave the second check

**Comment by ahsonkhan** [comment]
Not related to MQ :)

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp]: updated to OnRead(). I liked it

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: restored the second canceled() checked

**Comment by vhvb1989** [comment]
> Not related to MQ :)

Feel free to update this whenever you see it. It would be faster if you see it and act on it. I don't really care too much about the milestone records.  I don't really know if we really should add a milestone to the PRs, I just do it because you asked me for it once and I don't mind XD

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:77]:
```suggestion
     * @remark Checks and throws if context is cancelled. Otherwise, calls the `OnRead` method that must be overridden.
```

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:562]: Remove this.
```suggestion
```

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: @JeffreyRichter - can you help answer this? I think we should side towards checking the cancellation more frequently, before any thread sleeps, and not just before I/O operations.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp]: We should add this in the guidelines so we are consistent with `Operation<T>` and other usages of private/protected virtual methods. Either we do: `OnX` or `XImplementation`, so `OnPollUntilDone`?

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (JeffreyRichter) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:85]: It should not be virtual so we can enforce this behavior.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (JeffreyRichter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Yes, we can check before a thread sleeps too. We could actually test if the expiration will happen while sleeping and then NOT sleep and just throw.

Comment (vhvb1989) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: > Yes, we can check before a thread sleeps too.
@JeffreyRichter the retry policies do sleep... So, should we return the cancellation check to the retry policies?
Or should we just ignore this PR ?

Comment (JeffreyRichter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Yes, should have a check just before the sleep. As I said above, if the cancellation will happen while sleeping, then we should just throw and not sleep.

Comment (vhvb1989) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: So, would this mean that we can throw before the timeout is reached right?

like
```cpp
Context context(5minutesTimeOut);

// Then from retry policy
Time sleepFor = 10minutes;
context.ThrowIfCancellIn(sleepFor);  // This might throw super early before the 5 min
Thread.sleep(sleepFor);
```

Do we need to document this?  Or just is it enough to tell in the exception that The SDK has detected a future TimeOut and decided to throw in advance?

Comment (JeffreyRichter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Good point. It just makes no sense to sleep and then cancel. So, I think we should do this and do NOT think we need to document it - let's see if any customer even notices. But, if they do, we'll just document it.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Agreed. Checking and throwing if context is cancelled right before thread sleep is good enough.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/azure/core/context.hpp:419]: I suggest we NOT have this public function on Context and that just do this near where we sleep. This way, we don't have to support it, document it, etc. We just have to make it work in the 1 place where we need it.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (JeffreyRichter) [inline sdk/core/azure-core/src/http/retry_policy.cpp:180]: Just move the code inline here.

File: sdk/core/azure-core/test/ut/context.cpp
Comment (JeffreyRichter) [inline sdk/core/azure-core/test/ut/context.cpp:104]: You can have a private or internal function in order to unit test (if you desire).

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (JeffreyRichter) [inline sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:125]: OK, make the function internal (not private) and call it in both retry policies.

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:125]: Sure.

Question:  We can make this function to behave in 2 ways:

Imagine this case:  
 - Create context `C` with 100ms time out.
 - Send request, then reach to a point where we are going to sleep for 500 ms.

First result: Since we know the context will throw, we just do it early and before reaching 100ms.  (this is the actual implementation)

Second result: Sleep for exactly the time left for cancelling the request.  `ThrowIfWillCancelledAfter` will call sleep for any time left to only reach the time out and then it throws. This way we wouldn't throw early and we won't also wait for more than the requested time (when the sleep time is greater than that, like this example)

**Comment by JeffreyRichter** [comment]
I would keep current behavior. It feels very bad to me to sleep just to throw.





Jeffrey Richter (watch Architecting Distributed Cloud Apps)

________________________________

**Review by katmsft — APPROVED** [review]
The storage side of the changes looks good.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/context.hpp:419]: Are we going with 1 'l' or with 2?  I see we have `IsCanceled()`, `ThrowIfCanceled()` and now `ThrowIfWillCancelledAfter()`.  

The name of this method is odd?  `ThrowIfWillCancelledAfter()`...  `ThrowIfWillCancelAfter` reads better.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:156]: Why is OnRead public?  As a customer I expect to call `Read`

**Review by RickWinter — CHANGES_REQUESTED** [review]
We guarantee that we will cancel early, but that is what we are doing here.  We are assuming the network call will be longer than the interval.  I don't agree that is always true

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (RickWinter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: I don't agree with this policy.  Our design is we guarantee we won't be early but we don't guarantee cancellation happens on time.  This breaks that design.  

Also cancelling before a sleep isn't a good idea.  There are valid programs which sleep a thread and then call resume on the sleeping threads handle from another thread.  Resume decrements the suspend count and will result in the sleeping thread waking up early.

**Review by Jinming-Hu — APPROVED** [review]
Changes to storage code looks good to me.

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/context.hpp:419]: with 2.  But updating all the existing occurrences is part of another PR #1256

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:156]: Very good catch!
It should be protected, and I need to update this to all concrete BodyStream classes. THank you

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: @RickWinter we would only call the `Azure::Core::Internal::ThrowIfWillCancelAfter(ctx, retryAfter);` when we know the exact time that the thread will sleep and when we know in advance that there's no other mechanism to wake up and resume that thread.

For the case you describe, we shouldn't be checking for cancellation before sleeping.

I would also prefer to do the cancellation later than earlier, even if we sleep for a while just to cancel the request after that.  Or even if we don't even worry about cancelling before sleeping (since cancelling is just best effort).

@JeffreyRichter  @ahsonkhan  would you agree with Rick and me?  or do we want more conversation about it?

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:125]: Moved to be Internal API

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (JeffreyRichter) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: First, I don't think whatever we decide is a big deal either way. We can err on the safer side today by just sleeping as we normally would and then throw if canceled. This was we never cancel early but we block a thread for no good reason (which is less efficient but is not wrong).

If any customers have issue with this, then we could shrink the sleep time to the sooner of it & cancelation. 
If any customer have issue with this, then we could avoid the sleep entirely and cancel early.
My point being that we can wait for customer feedback on this and just do the simple implementation for now.

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: I agree with what @RickWinter suggested and do the simpler thing of checking once before sleeping if context has been cancelled, and that's it. Trying to calculate the subsequent sleep time and pre-emptively cancelling is error prone and unexpected, even though there is a non-optimal behavior. Ideally, we use thread sleep sparingly and for really small durations.

As @JeffreyRichter stated:
> and just do the simple implementation for now.

To clarify about these:
> I would also prefer to do the cancellation later than earlier, even if we sleep for a while just to cancel the request after that.

>  We can err on the safer side today by just sleeping as we normally would and then throw if canceled.

I agree, but check for cancellation and throw **before** sleeping. We can also check after, if desired, but let's **not** only check after.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:52]: Make this `private`, or is there a reason to keep it `protected`?
```suggestion
  private:
    /**
     * @brief Read portion of data into a buffer.
     *
     * @remark This is the Read implementation that all derived classes need to provide.
     *
     * @param context #Context so that operation can be canceled.
     * @param buffer Pointer to a byte buffer to read the data into.
     * @param count Size of the buffer to read the data into.
     *
     * @return Number of bytes read.
     */
    virtual int64_t OnRead(Context const& context, uint8_t* buffer, int64_t count) = 0;
```

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl_session_private.hpp:347]:
```suggestion
     * @brief Implement #BodyStream `OnRead`. Calling this function pulls data from the wire.
```

File: sdk/core/azure-core/inc/azure/core/http/winhttp/win_http_client.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/winhttp/win_http_client.hpp:89]:
```suggestion
       * @brief Implement #BodyStream `OnRead`. Calling this function pulls data from the wire.
```

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:44]:
```suggestion
     * @remark This is the `OnRead` implementation that all derived classes need to provide.
```

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/retry_policy.cpp:180]:
```suggestion
      ctx.ThrowIfCanceled();
```

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:126]:
```suggestion
          ctx.ThrowIfCanceled();
```

File: sdk/core/azure-core/inc/azure/core/context.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/context.hpp:407]: In fact, if we are removing `ThrowIfWillCancelAfter`, revert the changes to this whole `context.hpp` file.

File: sdk/core/azure-core/src/http/retry_policy.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/retry_policy.cpp:186]: Reverted the feature to check if will be cancelled. :)

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:127]:
```suggestion
          std::this_thread::sleep_for(std::chrono::milliseconds(retryDelayMs));
```

File: sdk/storage/azure-storage-common/src/storage_retry_policy.cpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-common/src/storage_retry_policy.cpp:125]:
```suggestion
```

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp
Comment (RickWinter) [inline sdk/core/azure-core/src/http/curl/curl_session_private.hpp:353]: This should be private.

**Review by RickWinter — APPROVED** [review]
ResponseBufferParser needs OnRead moved to private before submitting.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:40]: Do we need this to be protected, or can it remain private?
```suggestion
  private:
```

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl_session_private.hpp:353]: It makes sense to be explicit and add the `private:` keyword, but just to confirm, isn't this already private given defaults in C++ classes is private, and this is written before the `public:` keyword?

**Review by ahsonkhan — APPROVED** [review]
Mark `BodyStream::OnRead` as private instead of protected on the base type. Otherwise, looks good.

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl_session_private.hpp:353]: Yes, it is already private.
The class declaration before it makes it a little confusing. But this is a private member next to a private class definition

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl_session_private.hpp:353]: > The class declaration before it makes it a little confusing. But this is a private member next to a private class definition

Maybe we can move things around a bit to make it clearer. Can we move the nested class at the beginning or end, rather than declaring a class in the middle of a list of public members for the outer class?

**Pull request merged**
